nodejs unit tests and documentation links

davidchaiken opened the pull request:
* added lots of nodejs unit tests
* unified api_object for v3 and v5 (similar to work done in python in the last pull request)
* refactored a few lines of code (e.g. in api_common.js) to get the tests to work
* added missing references to endpoint documentation

Note: Due to the way that I branched on my laptop, a number of commits listed for this PR were already merged in pull request #4. The diffs in the Files changed tab look correct even though the commit list is confusing.

@thisjuvy

thisjuvy commented on nodejs/src/access_token_common.js:
```diff
@@ -7,11 +7,10 @@ import {ApiCommon} from './api_common.js'
 export class AccessTokenCommon extends ApiCommon {
 
   constructor(api_config, {name = null}) {
```
@davidchaiken this could simply be {name} as it will be undefined if not found in the input object.

thisjuvy replied:
Unless null is a valid assignment when not provided, then you should keep it as is.

davidchaiken replied:
I removed all of the instances of argument initialization to null, except where null is a valid assignment (in the case of v3 scopes).

davidchaiken replied:
Oops! This change didn't work out for me. I'm getting this error:
```TypeError: Cannot destructure property 'name' of 'undefined' as it is undefined.```

In the cases where I specified null, null is a valid assignment when not provided.

thisjuvy commented on nodejs/src/access_token_common.js:
```diff
@@ -7,11 +7,10 @@ import {ApiCommon} from './api_common.js'
-    this.api_config = api_config;
     this.api_uri = api_config.api_uri;
     this.auth_headers = {'Authorization': 'Basic ' + b64auth}
     if (name) {
```
You can use short hand if statement here and a template literal for the string concat. We should probably stick to a convention. Either use "somestring "+somevar; or `somestring ${somevar}`;

this.name = (name) ? name: `access_token_${api_config.version}`;

davidchaiken replied:
I switched the code to templated strings in as many places as make sense, and used the trinary operator to set access_token.name.

thisjuvy commented on nodejs/src/access_token_common.js:
```diff
@@ -100,8 +99,8 @@ export class AccessTokenCommon extends ApiCommon {
                     'access_token': this.access_token,
```
nit: you could do

```
{
  name: this.name,
  access_token: this.access_token,
  refresh_token: this.refresh_token,
  scopes: this.scopes
};
```

any maybe do without the intermediate output variable? simply pass it as input, I don't see output being used elsewhere in this method.

davidchaiken replied:
done. also modified other structures in the repo to eliminate quotes on keys.

thisjuvy commented on nodejs/src/api_object.js:
```diff
@@ -1,19 +1,28 @@
+  // This method extracts the data container from the v3 response body
+  // and returns the v5 response body without modification.
+  extract(body) {
+    if (this.api_config.version == 'v3') {
```
`return (this.api_config.version === 'v3') ? body.data : data;`

davidchaiken replied:
done.

thisjuvy reviewed (commented):
left comments. Mostly stylistic feedback.

davidchaiken commented:
> left comments. Mostly stylistic feedback.

Thanks, @thisjuvy! I think that I've addressed all of your comments. Looking for more comments or a +1...

thisjuvy commented on nodejs/src/api_object.js:
```diff
@@ -1,19 +1,24 @@
+  // This method extracts the data container from the v3 response body
+  // and returns the v5 response body without modification.
+  extract(body) {
+    return (this.api_config.version == 'v3') ? body.data : body;
```
nit: use === as we don't need type conversion before checking equality.

davidchaiken replied:
done. changed all instances of == to === because type conversion is not necessary for any of the comparisons.
